File: internal/types/policy.go
package types

import (
	"encoding/json"
)

type PluginTriggerEvent struct {
	PolicyID string `json:"policy_id"`
}

type PluginPolicy struct {
	ID             string          `json:"id" validate:"required"`
	PublicKeyEcdsa string          `json:"public_key_ecdsa" validate:"required"`
	PublicKeyEddsa string          `json:"public_key_eddsa" validate:"required"`
	PluginVersion  string          `json:"plugin_version" validate:"required"`
	PolicyVersion  string          `json:"policy_version" validate:"required"`
	PluginType     string          `json:"plugin_type" validate:"required"`
	IsEcdsa        bool            `json:"is_ecdsa"`
	ChainCodeHex   string          `json:"chain_code_hex" validate:"required"`
	DerivePath     string          `json:"derive_path" validate:"required"`
	Active         bool            `json:"active"`
	Progress       string          `json:"progress" validate:"required"`
	Signature      string          `json:"signature" validate:"required"`
	Policy         json.RawMessage `json:"policy" validate:"required"`
}

type PluginPolicyPaginatedList struct {
	Policies   []PluginPolicy `json:"policies" validate:"required"`
	TotalCount int            `json:"total_count" validate:"required"`
}

/**
 * Returns public key to sign transactions with, based on the chain for which the policy is signed
 */
func (p *PluginPolicy) GetPublicKey() string {
	if p.IsEcdsa {
		return p.PublicKeyEcdsa
	}
	return p.PublicKeyEddsa
}

File: plugins-ui/src/modules/policy/models/policy.ts
import { RJSFSchema, UiSchema } from "@rjsf/utils";
import { ColumnDef } from "@tanstack/react-table";

export enum PluginProgress {
  InProgress = 'IN PROGRESS',
  Done = 'DONE',
};

export type Policy<
  T = string | number | boolean | string[] | null | undefined,
> = {
  [key: string]: T | Policy<T>;
};

export type PluginPolicy = {
  id: string;
  public_key_ecdsa: string;
  public_key_eddsa: string;
  plugin_version: string;
  policy_version: string;
  plugin_type: string;
  is_ecdsa: boolean;
  chain_code_hex: string;
  derive_path: string;
  active: boolean;
  progress: PluginProgress;
  signature: string;
  policy: Policy;
};

export type PluginPoliciesMap = {
  policies: PluginPolicy[];
  total_count: number;
};

export type TransactionHistory = {
  history: PolicyTransactionHistory[];
  total_count: number;
};

export type PolicyTableColumn = ColumnDef<unknown> & {
  accessorKey: string;
  header: string;
  cellComponent?: string;
  expandable?: boolean;
};

export type PolicyTransactionHistory = {
  id: string;
  updated_at: string;
  status: string;
};

export type PolicySchema = {
  form: {
    schema: RJSFSchema;
    uiSchema: UiSchema;
    plugin_version: string;
    policy_version: string;
    plugin_type: string;
  };
  table: {
    columns: PolicyTableColumn[];
    mapping: Record<string, string | string[]>;
  };
};
